### PR TITLE
[BUG] `FallbackForecaster` fails with `ForecastByLevel` when `nan_predict_policy='raise'`

### PR DESCRIPTION
**Pull request by ninedigits**
Fixes error that occurs when `FallbackForecaster` is combined with `ForecastByLevel` and `nan_predict_policy='raise'`
https://github.com/sktime/sktime/issues/6230

### CODE REVIEW

File: CONTRIBUTORS.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ Contributors
 <!-- Please add your badges to .all-contibutorsrc -->
 
 <!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
-[![All Contributors](https://img.shields.io/badge/all_contributors-266-orange.svg)](#contributors)
+[![All Contributors](https://img.shields.io/badge/all_contributors-267-orange.svg)](#contributors)
```
Comment (ninedigits) [inline CONTRIBUTORS.md:9]: Looks like this change got incorporated that I wasn't aware. @fkiraly should I resubmit this PR?

Comment (fkiraly) [reply CONTRIBUTORS.md:9]: no, that's automatic, it's fine

**Comment by ninedigits** [comment]
None of the original tests were designed to handle MultiIndex DataFrames. In scenarios where `y_pred` is typically a Series, it functions as expected; however, it becomes a DataFrame when the input series has a MultiIndex structure.

In the first case, where `y_train` is a DataFrame with a single index, calling `y_pred.isna().any()` returns a boolean, which works fine for conditional checks. However, in the second case, where `y_train` has a MultiIndex, as in with hierarchical dataframes, the same call results in a Series instead of a single boolean value.  This causes an exception because boolean logic operations can't be directly applied to a Series.

**Comment by ninedigits** [comment]
On a side note, I see a notification for changes requested but I don't see any actual change requests to the code. Am I missing something?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Oh, I see! I understand now.

The logic attached to `nan_predict_policy` required `pd.DataFrame` format.

Then this indicates a bigger, unreported bug, which we should try to fix, let me explain.

For `y_inner_mtype`, `X_inner_mtype`, we allowed `ALL_TIME_SERIES_MTYPES`, because the exception handling logic did not require any specific type. E.g., 3D `numpy` arrays, weird nested `pd.DataFrame`, and `dask` array were all fine.

When you added the nan handling policy, you assumed `pd.DataFrame`, and this was fine since the regular testing scenarios only resulted in few types to be passed.

There are a few ways to fix the issue:

1. if `nan_handling_policy` is nof `ignore`, in `__init__`, we set the `X_inner_mtype` and `y_inner_mtype` via `self.set_tags` to the types supported, presumbaly `["pd.DataFrame", "pd-multiindex", "pd_multiindex_hier"]`. This ensures that in `_predict` and `_fit`, you only ever see these types.
2. or, leave the types as they are, and ensure the logic is correct for all data types.

I would say that 2. is infeasible, so we ought to do 1.?

Regarding test coverage, this shows a gap in our testing framework - estimators aren't tested with all data types they could get.

Are you certain that the hierarchical type causes the issues? Then we may have to add a forecaster tests with a hierarchical example. I thought this was covered.

> On a side note, I see a notification for changes requested but I don't see any actual change requests to the code. Am I missing something?

No, it is possible to make change requests without reference to code lines, e.g., in general questions like these.

**Comment by ninedigits** [comment]
> Are you certain that the hierarchical type causes the issues? Then we may have to add a forecaster tests with a hierarchical example. I thought this was covered.

Pretty sure, I'm attaching a screenshot of the debug window to highlight:

Screenshot 2024-04-02 at 2 50 09 PM

When working with hierarchical data (in the form of a dataframe) using `ForecastByLevel`, the function `forecaster.predict()` will produce a dataframe as the output. However, when forecasting a series that includes an input series, `y_pred` will be returned as a series.

The original error was that the boolean logic I was applying to determine the presence of nans works differently on dataframes than series. When you say:

> Then this indicates a bigger, unreported bug, which we should try to fix, let me explain.

Do you mean with `FallbackForecaster` or with other estimators in general?

**Comment by ninedigits** [comment]
> Only with `FallbackForecaster`, and a possible solution is discussed above.
> 
> >

Unless I'm misunderstanding something here I think the error is a pretty simple fix.

This test reproduces the error:
```python3
def test_forecastbylevel_nan_predict():
    from sktime.forecasting.compose import ForecastByLevel
    from sktime.utils._testing.hierarchical import _make_hierarchical

    df = _make_hierarchical(
        hierarchy_levels=(2, 2, 2),
        max_timepoints=10,
        min_timepoints=10,
        same_cutoff=True,
        n_columns=1,
        all_positive=True,
        index_type="period",
        random_state=0,
        add_nan=False,
    )
    forecaster1 = DummyForecaster(raise_at=None, predict_nans=True)
    forecaster2 = NaiveForecaster()
    forecaster = ForecastByLevel(
        FallbackForecaster(
            [
                ("forecaster1_pred_nans", forecaster1),
                ("forecaster2_expected_", forecaster2),
            ],
            nan_predict_policy="raise",
        )
    )
    fh = [1, 2, 3]
    forecaster.fit(y=df, fh=fh)
    y_pred_actual = forecaster.predict()

    forecaster2.fit(y=df, fh=[1, 2, 3])
    y_pred_expected = forecaster2.predict()

    pd.testing.assert_frame_equal(y_pred_expected, y_pred_actual)
```

Error:
```stdout
>           self._validate_y_pred(y_pred)

X          = None
fh         = ForecastingHorizon([1, 2, 3], dtype='int64', is_relative=True)
self       = FallbackForecaster(forecasters=[('forecaster1_pred_nans',
                                 DummyForecaster(predict_nan...                          ('forecaster2_expected_', NaiveForecaster())],
                   nan_predict_policy='raise')
y_pred     =                c0
2000-11       NaN
2000-12  3.963588
2001-01  3.963588

../_fallback.py:269: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../_fallback.py:160: in _validate_y_pred
    if has_nans:
        has_nans   = c0    True
dtype: bool
        self       = FallbackForecaster(forecasters=[('forecaster1_pred_nans',
                                 DummyForecaster(predict_nan...                          ('forecaster2_expected_', NaiveForecaster())],
                   nan_predict_policy='raise')
        y_pred     =                c0
2000-11       NaN
2000-12  3.963588
2001-01  3.963588
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = c0    True
dtype: bool

    @final
    def __nonzero__(self) -> NoReturn:
>       raise ValueError(
            f"The truth value of a {type(self).__name__} is ambiguous. "
            "Use a.empty, a.bool(), a.item(), a.any() or a.all()."
        )
E       ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

**Comment by fkiraly** [comment]
> Unless I'm misunderstanding something here I think the error is a pretty simple fix.

I think you are, the issue is that users can pass any of 10 data formats to forecasters, and they are converted if not internally supported. This PR fixes the issue for one additional format; the programmatic way would be to set the tag.

**Comment by fkiraly** [comment]
Here's another option: instead of the custom nan checking logic, we could use `check_is_mtype` against the expected `mtype`, and request only the `has_nans` property.

The expected mtype is in `self._y_mtype_last_seen`.

This might be the simplest way to fix this?

**Comment by fkiraly** [comment]
Please let me know if this sounds confusing and I can try to add the fix. It's a few lines, given the discussion.

**Comment by fkiraly** [comment]
> If we were to set the mtypes in init, as in your initial suggestion, would it look something like this?

Yes, that's how it would look like.

I think my last solution is better though, using `check_is_mtype` for the "has nan" check, that covers all data types and avoids conversions.

**Comment by fkiraly** [comment]
> the code still executes when using hierarchical data.

which code did you expect not to execute?

**Comment by ninedigits** [comment]
> > the code still executes when using hierarchical data.
> 
> which code did you expect not to execute?

I think I understand now. I was assuming that somewhere in the code, an exception would be raised if the wrong data type was passed in, eg within `BaseForecaster`. So I was playing around with `set_tag` with the different datatypes, making it more restrictive to see if exceptions were raised. But I'm looking through the code now and I think what you're getting at is that we'd have an additional check within `validate_y_pred` and the exception would need to be programmed there. That way, as the system evolves, and new data types are passed in, it points to that specific area to review and add those test cases in. Is that right?

**Comment by fkiraly** [comment]
yes, exactly.
Would you like me to add the logic with `check_is_mtype`, or would you like to give it a try?

**Comment by ninedigits** [comment]
> yes, exactly. Would you like me to add the logic with `check_is_mtype`, or would you like to give it a try?

I think I get it. Would it look something like this?

```python3
    def _validate_y_pred(self, y_pred):
        if self.nan_predict_policy in ("warn", "raise"):
            last_mtype = self._y_mtype_last_seen
            expected_mtypes = ["pd.DataFrame", "pd-multiindex", "pd_multiindex_hier"]
            passed_mcheck = check_is_mtype(
                y_pred,
                expected_mtypes
            )
            if not passed_mcheck:
                msg = "`nan_predict_policy` expects mtype data to be one of " \
                      f"{expected_mtypes} but intead got {last_mtype}. If you think " \
                      f"FallbackForecaster's nan_predict_policy should be able to " \
                      f"handle this datatype, raise an issue on sktime"
                raise NotImplementedError(msg)
```

I'm not sure what's the best way to add a test case though. What other types of data are available to forecast with? I didn't see any obvious resource online to read more about it.

**Comment by fkiraly** [comment]
not exactly, I was thinking to use the check to get the metadata field `has_nans`, i.e., 

```python3
    def _validate_y_pred(self, y_pred):
        if self.nan_predict_policy in ("warn", "raise"):
            last_mtype = self._y_mtype_last_seen
            _, _, metadata = check_is_mtype(
                y_pred,
                mtype=last_mtype,
                return_metadata=["has_nans"],
            )
            has_nans = metadata["has_nans"]
```

Now you have a variable `has_nans` that tells you - for any data container that you might see, `pd.DataFrame`, `pd.Series`, or anything exotic, whether the container has NA.

Noting, we *know* that the mtpe should be `_y_mtype_last_seen`, if the base class interface is not buggy (because "same mtype" is guaranteed).
We are "abusing" the data checker to get the metadata which we are interested in.

**Comment by ninedigits** [comment]
So I tried that and it's failed the below unit test:

```python3
def test_fallbackforecaster_predict_nan():
    """Test FallbackForecaster raise if nans in predict"""
    y = make_forecasting_problem(random_state=42)
    forecaster1 = DummyForecaster(raise_at="fit")
    forecaster2 = DummyForecaster(raise_at=None, predict_nans=True)
    forecaster3 = PolynomialTrendForecaster()
    forecaster4 = NaiveForecaster()

    forecaster = FallbackForecaster(
        [
            ("forecaster1_fails_fit", forecaster1),
            ("forecaster2_pred_nans", forecaster2),
            ("forecaster3_succeeded", forecaster3),
            ("forecaster4_notcalled", forecaster4),
        ],
        nan_predict_policy="raise",
    )
    forecaster.fit(y=y, fh=[1, 2, 3])
    y_pred_actual = forecaster.predict()

    forecaster3.fit(y=y, fh=[1, 2, 3])
    y_pred_expected = forecaster3.predict()

    # Assert correct forecaster name
    name = forecaster.current_name_
    assert name == "forecaster3_succeeded"

    # Assert correct y_pred
    pd.testing.assert_series_equal(y_pred_expected, y_pred_actual)

    # Assert correct number of expected exceptions
    exceptions_raised = forecaster.exceptions_raised_
    assert len(exceptions_raised) == 2

    # Assert the correct forecasters failed
    names_raised_actual = [
        vals["forecaster_name"] for vals in exceptions_raised.values()
    ]
    names_raised_expected = ["forecaster1_fails_fit", "forecaster2_pred_nans"]
    assert names_raised_actual == names_raised_expected
```
Looks like the training data is a pandas series and the prediction is a pandas dataframe:

Screenshot 2024-04-04 at 12 51 22 PM

**Comment by fkiraly** [comment]
perhaps - could you kindly check, which forecaster returns the unexpected type?

**Comment by ninedigits** [comment]
> perhaps - could you kindly check, which forecaster returns the unexpected type?

I think there's a bug either in `DummyForecaster` or `FallbackForecaster`. I have some time to investigate today, will update this comment when I have more information.

**Comment by fkiraly** [comment]
Ah, of course, that was it - the two fixes should not be applied at the same time.

> @fkiraly, would it be appropriate to remove this line and maintain the logic as described in your suggestion, i.e., by adding the `check_mtype` in the validation check method?

Yes, that is imo indeed the fix to the secondary bug - simply remove any additional logic from `__init__`.

**Comment by ninedigits** [comment]
@fkiraly Any more updates needed on my end or are we good to go? I saw that the check failed earlier but I wasn't sure if it was an actual issue with the code.

**Comment by fkiraly** [comment]
let's see - I'll restart the tests

**Review by fkiraly — APPROVED** [review]
👍

**Pull request merged**